pkg/storage/prunable/epochstore/store.go

```diff
@@ -136,40 +153,188 @@ func (s *Store[V]) Prune(epoch iotago.EpochIndex, defaultPruningDelay iotago.Epo
+			return prunedEpochs, ierrors.Wrapf(err, "failed to store lastPrunedEpoch for epoch %d in Prune", i)
 		}
+
+		prunedEpochs = append(prunedEpochs, i)
 	}
 
-	return nil
+	return prunedEpochs, nil
 }
 
-func (s *Store[V]) RollbackEpochs(epoch iotago.EpochIndex) (lastPrunedEpoch iotago.EpochIndex, err error) {
+func (s *BaseStore[V]) RollbackEpochs(epoch iotago.EpochIndex) (lastPrunedEpoch iotago.EpochIndex, rollbackEpochs []iotago.EpochIndex, err error) {
+	rollbackEpochs = make([]iotago.EpochIndex, 0)
+
 	lastAccessedEpoch, err := s.LastAccessedEpoch()
 	if err != nil {
-		return lastAccessedEpoch, ierrors.Wrap(err, "failed to get last accessed epoch")
+		return lastAccessedEpoch, rollbackEpochs, ierrors.Wrap(err, "failed to get last accessed epoch")
 	}
 
 	for epochToPrune := epoch; epochToPrune <= lastAccessedEpoch; epochToPrune++ {
 		if err = s.DeleteEpoch(epochToPrune); err != nil {
-			return epochToPrune, ierrors.Wrapf(err, "error while deleting epoch %d", epochToPrune)
+			return epochToPrune, rollbackEpochs, ierrors.Wrapf(err, "error while deleting epoch %d", epochToPrune)
 		}
+
+		rollbackEpochs = append(rollbackEpochs, epochToPrune)
 	}
 
-	return lastAccessedEpoch, nil
+	return lastAccessedEpoch, rollbackEpochs, nil
 }
 
-func (s *Store[V]) isTooOld(epoch iotago.EpochIndex) bool {
+func (s *BaseStore[V]) isTooOld(epoch iotago.EpochIndex) bool {
 	prunedEpoch, hasPruned := s.lastPrunedEpoch.Index()
 
 	return hasPruned && epoch <= prunedEpoch
 }
+
+type CachedStore[V any] struct {
+	store *BaseStore[V]
+
+	cache      *shrinkingmap.ShrinkingMap[iotago.EpochIndex, V]
+	addedQueue *queue.Queue[iotago.EpochIndex]
+
+	mutex syncutils.RWMutex
+}
+
+func NewCachedStore[V any](store *BaseStore[V], maxCacheSize int) *CachedStore[V] {
+	return &CachedStore[V]{
+		store:      store,
+		addedQueue: queue.New[iotago.EpochIndex](maxCacheSize),
+		cache:      shrinkingmap.New[iotago.EpochIndex, V](),
+	}
+}
+
+func (c *CachedStore[V]) addToCache(epoch iotago.EpochIndex, value V) {
+	if !c.addedQueue.Offer(epoch) {
+		removedEpoch, _ := c.addedQueue.Poll()
+		c.cache.Delete(removedEpoch)
+
+		c.addedQueue.Offer(epoch)
+	}
+
+	c.cache.Set(epoch, value)
+}
+
+func (c *CachedStore[V]) removeFromCache(epoch iotago.EpochIndex) {
+	// removing from addedQueue is not necessary, as it will be overwritten anyway when the cache gets full.
+	c.cache.Delete(epoch)
+}
+
+func (c *CachedStore[V]) RestoreLastPrunedEpoch() error {
+	return c.store.RestoreLastPrunedEpoch()
+}
+
+func (c *CachedStore[V]) LastAccessedEpoch() (lastAccessedEpoch iotago.EpochIndex, err error) {
+	return c.store.LastAccessedEpoch()
+}
+
+func (c *CachedStore[V]) LastPrunedEpoch() (iotago.EpochIndex, bool) {
+	return c.store.LastPrunedEpoch()
+}
+
+func (c *CachedStore[V]) Load(epoch iotago.EpochIndex) (V, error) {
+	var zeroValue V
+
+	c.mutex.RLock()
+	if value, exists := c.cache.Get(epoch); exists {
```

Comment:
This should be using Compute and loading it from store inside the func instead of having an extra mutex

Reply:
I'd avoid using compute as this will use a write lock in this hot path. It is ugly but imo worth the performance we get form it. I added some comments and an additional check whether it has been added in the meantime to the write locked part.